File: services/15-django/api/auth/jwt.py
"""JWT creation and verification — identical to 15-fastapi canonical."""

import os
from datetime import datetime, timedelta, timezone
from typing import Optional

from jose import JWTError, jwt

JWT_SECRET: str = os.getenv("JWT_SECRET", "change-me-in-production-use-at-least-32-chars")
JWT_ALGORITHM: str = "HS256"
JWT_EXPIRY_HOURS: int = 8


def create_access_token(user_id: int, email: str, name: str) -> str:
    """Issue a signed JWT for the given user (8-hour expiry)."""
    now = datetime.now(timezone.utc)
    payload: dict = {
        "sub": str(user_id),
        "email": email,
        "name": name,
        "iat": now,
        "exp": now + timedelta(hours=JWT_EXPIRY_HOURS),
    }
    return jwt.encode(payload, JWT_SECRET, algorithm=JWT_ALGORITHM)


def verify_token(token: str) -> Optional[dict]:
    """Decode and verify a JWT.

    Returns the decoded payload dict when valid.
    Returns None when the token is expired, malformed, or has a bad signature.
    """
    try:
        payload = jwt.decode(token, JWT_SECRET, algorithms=[JWT_ALGORITHM])
        return payload
    except JWTError:
        return None

File: services/30-ws-elixir/lib/ws_elixir/auth/jwt.ex
defmodule WsElixir.Auth.Jwt do
  @moduledoc """
  JWT creation and verification for WebSocket auth.

  Secret: loaded from JWT_SECRET env var.
  Expiry: 8 hours from issuance.
  Algorithm: HS256.

  Requires: joken (add {:joken, "~> 2.6"} to mix.exs deps)
  """

  @ttl_seconds 8 * 60 * 60

  @doc """
  Issues a signed JWT for the given user.

  Returns {:ok, token} on success.
  Returns {:error, reason} when JWT_SECRET is not set.
  """
  def sign_token(user_id, email, name) do
    with {:ok, secret} <- fetch_secret() do
      signer = Joken.Signer.create("HS256", secret)
      now = System.system_time(:second)

      claims = %{
        "sub"   => to_string(user_id),
        "email" => email,
        "name"  => name,
        "iat"   => now,
        "exp"   => now + @ttl_seconds
      }

      Joken.encode_and_sign(claims, signer)
    end
  end

  @doc """
  Verifies and decodes a JWT string.

  Returns {:ok, claims} when the token is valid and not expired.
  Returns {:error, reason} when the token is invalid, expired, or the secret is unset.
  """
  def verify_token(token) do
    with {:ok, secret} <- fetch_secret() do
      signer = Joken.Signer.create("HS256", secret)
      Joken.verify_and_validate(token, signer)
    end
  end

  defp fetch_secret do
    case System.get_env("JWT_SECRET") do
      nil -> {:error, "JWT_SECRET environment variable is not set"}
      ""  -> {:error, "JWT_SECRET environment variable is not set"}
      secret -> {:ok, secret}
    end
  end
end
